### PR TITLE
Implement UNION, ??, and IF/ELSE as generic operators

### PR DESCRIPTION
**Pull request by elprans**
See individual commits.

### CODE REVIEW

File: edb/edgeql/compiler/func.py [hunk 1]
```diff
@@ -468,8 +485,15 @@ def finalize_args(bound_call: polyres.BoundCall, *,
...
         val_material_type = barg.valtype.material_type(ctx.env.schema)
         param_material_type = paramtype.material_type(ctx.env.schema)
-        if not val_material_type.issubclass(
-                ctx.env.schema, param_material_type):
+
+        compatible = (
+            val_material_type.issubclass(ctx.env.schema, param_material_type)
+            and (not param_material_type.is_tuple()
+                 or (param_material_type.get_element_names() ==
+                     val_material_type.get_element_names()))
```
Comment (1st1) [inline edb/edgeql/compiler/func.py]: `element_types`?

File: edb/edgeql/parser/grammar/ddl.py [hunk 2]
```diff
@@ -1775,6 +1775,9 @@ def reduce_POSTFIX(self, *kids):
     def reduce_PREFIX(self, *kids):
         self.val = kids[0].val
 
+    def reduce_NARY(self, *kids):
```
Comment (1st1) [inline edb/edgeql/parser/grammar/ddl.py]: ternary?

Comment (1st1) [reply edb/edgeql/parser/grammar/ddl.py]: We can add N-ary when we need to, but if/else it ternary, right?

File: tests/test_edgeql_syntax.py [hunk 3]
```diff
@@ -2368,10 +2368,10 @@ def test_edgeql_syntax_updatefor_01(self):
 
     def test_edgeql_syntax_coalesce_01(self):
         """
-        SELECT a ?? x;
-        SELECT a ?? x.a;
-        SELECT a ?? x.a[IS ABC];
-        SELECT (a ?? x.a[IS ABC]@aaa + 1);
+        SELECT (a ?? x);
```
Comment (1st1) [inline tests/test_edgeql_syntax.py]: Why these changes?

File: edb/edgeql/parser/grammar/ddl.py (hunk 2 above)
Comment (elprans) [reply edb/edgeql/parser/grammar/ddl.py]: Yes, but ternary and n-ary are the same from the standpoint of handling them, so it makes little sense in having a separate kind for every value of N.

Comment (1st1) [reply edb/edgeql/parser/grammar/ddl.py]: Arent' binary and ternary and n-ary all the same from the standpoint of handling them? ;)

File: edb/edgeql/compiler/func.py (hunk 1 above)
Comment (1st1) [reply edb/edgeql/compiler/func.py]: Can you add a comment?
